0010-keccak-syscall.md

```diff
@@ -0,0 +1,402 @@
+168[`. While reading the preimage data, we can fill at the same time the sponge
+state. When 136 bytes have been read, a keccak permutation is executed.
+If we have finished reading the data, we run the permutation. In the case the total
+length is a multiple of 136, an additional permutation is executed
+with the expected padding.
+
+Reading additional bytes after the EOF must not have any effect and must be
+allowed. This can be achieved by activating the "preimage read gate" iff the
+different between the number of bytes that have to be read and the preimage
+offset is not zero. In the case we arrived at the end of the file, the preimage
+offset must not be increased and the number of bytes read must be set `0`.
+
+After we read the data and run the permutation a last time as described above,
+we must verify that the preimage key corresponds to the output of sponge.
+To achieve this, we compare the 31 bytes first bytes as the first byte is
+dropped for the preimage key type. The output of the keccak sponge is described
+in the [Keccak RFC](./keccak.md).
+
+
+#### Hint response
+
+<!-- DONE -->
+
+When `a0` equals `3` (alias by `fdHintRead`), we simulate we read all data by
+setting `v0` to the value contained in the register `a2`. It has no other effect.
+
+#### Default case
+
+<!-- DONE -->
+
+In the case of any other value for the input channel, the register `v0` must be
+set to `0xFFffFFff` and the error code `0x9` must be written in the register
+`a3`.
+
+
+### `sysWrite`
+
+This syscall is used to write data into an output channel specified by a file descriptor.
+The file descriptor is in the register `a0`. The address of the data is in the
+register `a1` and the length in bytes of the data that must be written is in the
+register `a2`.
+At the end, the actual number of bytes that have been written must be set into
+the register `v0` and the error code, if any, must be written in the register
+`a3`.
+
+Here the different output channels that must be supported:
+
+#### Standard output
+
+<!-- DONE -->
+
+The standard output has the file descriptor value `1`.
+
+`a2` bytes must be read from `a1` and copied in the standard output.
+The number of bytes, i.e. the value in `a2` must be copied into `v0` and the
+value `0` must be set in `a3`.
+
+#### Standard error
+
+<!-- DONE -->
+
+The standard error has the file descriptor value `2`.
+`a2` bytes must be read from `a1` and copied in the standard error.
+
+The number of bytes, i.e. the value in `a2` must be copied into `v0` and the
+value `0` must be set in `a3`.
+
+#### Hint write
+
+<!-- DONE -->
+
+In Cannon, even though the body is not empty, the code has no effect on the
+output. It can therefore be ignored.
+The number of bytes, i.e. the value in `a2`, must be copied into `v0` and
+the value `0` must be set in `a3`.
+
+#### Pre-image Oracle
+
+<!-- DONE -->
+
+The file descriptor `fdPreimageWrite` (value `6`) is used to read a pre-image
+key from the memory.
```

Comment:
when is this memory loaded with data so you can later read from it? is it the CLI?